src/crimson/osd/pg.cc

```diff
@@ -851,4 +859,470 @@ void PG::handle_rep_op_reply(crimson::net::Connection* conn,
   backend->got_rep_op_reply(m);
 }
 
+crimson::osd::blocking_future<bool> PG::start_recovery_ops(size_t max_to_start)
+{
+  ceph_assert(is_primary());
```

Comment:
i'd suggest use `ceph_assert()` only for detecting errors from runtime instead of using it for catching logic error in implementation.